boudicca.base/publisher-event-ical/src/main/kotlin/base/boudicca/publisher/event/ical/CalendarService.kt

```diff
@@ -32,56 +30,56 @@ class CalendarService @Autowired constructor(@Value("\${boudicca.search.url}") p
-    private fun parseEndDate(endDate: String?): OffsetDateTime? {
-        if (endDate == null) {
-            return null
+    fun createEvent(
+        event: StructuredEvent
+    ): VEvent {
+        val vEvent = VEvent()
+        vEvent.setSummary(event.name)
+        vEvent.dateStart = DateStart(event.startDate.toDate())
+        vEvent.uid = Uid("event-${event.startDate}-${event.name}")
+
+        val endDate = event.getProperty(SemanticKeys.ENDDATE_PROPERTY).firstOrNull()
+        val locationName = event.getProperty(SemanticKeys.LOCATION_NAME_PROPERTY).firstOrNull()
+        val locationUrl = event.getProperty(SemanticKeys.LOCATION_URL_PROPERTY).firstOrNull()
+        val description = event.getProperty(SemanticKeys.DESCRIPTION_TEXT_PROPERTY).firstOrNull()
+        val url = event.getProperty(SemanticKeys.URL_PROPERTY).firstOrNull()
+
+        endDate?.let { (_, endDate) ->
+            vEvent.dateEnd = DateEnd(endDate.toDate())
+        }
+        buildTextWithUrlSuffix(locationName, locationUrl)?.let {
+            vEvent.location = Location(it)
+        }
+        buildTextWithUrlSuffix(description, url)?.let {
+            vEvent.description = Description(it)
+        }
+        url?.let { (_, url) ->
+            vEvent.url = Url(url.toString())
         }
 
-        return OffsetDateTime.parse(endDate, DateTimeFormatter.ISO_DATE_TIME)
+        return vEvent
     }
 
-    fun createEvent(
-        title: String,
-        startDateTime: OffsetDateTime,
-        location: String?,
-        endDateTime: OffsetDateTime?,
-        sequence: Int,
-    ): VEvent {
-        //todo get description?
-        val titleHash = title.hashCode()
-        val event = VEvent()
-        event.setSummary(title)
-        event.dateStart = DateStart(Date(startDateTime.toInstant().toEpochMilli()))
-        if (endDateTime != null) {
-            event.dateEnd = DateEnd(Date(endDateTime.toInstant().toEpochMilli()))
-            event.uid = Uid("event-${startDateTime}-${endDateTime}-${titleHash}")
+    private fun buildTextWithUrlSuffix(text: Pair<Key, String>?, url: Pair<Key, URI>?): String? {
+        val textText = text?.second
+        val urlText = url?.second?.toString()
+        return if (!textText.isNullOrBlank() && !urlText.isNullOrBlank()) {
+            "$textText ($urlText)"
         } else {
-            event.uid = Uid("event-${startDateTime}-${titleHash}")
+            return textText ?: urlText
         }
-
-        // set the event properties
-        if (location != null) {
-            event.location = Location(location)
-        }
-        event.sequence = Sequence(sequence)
-
-        return event
     }
 
     fun getEvents(query: String): ByteArray {
-        val events = searchClient.queryEvents(QueryDTO(query,0, Int.MAX_VALUE))
+        val events = searchClient.queryEvents(QueryDTO(query, 0, Int.MAX_VALUE))
         return createCalendar(events.result)
     }
 }
+
+private fun OffsetDateTime.toDate() = Date.from(this.toInstant())
```

Comment:
this could go into a helper

Reply:
not sure, it's just a small line and there is no other class in the ical-project. not sure if this is important enough to put into the common-model, which is the only other place we could put it